pkg/reconciler/workload/apibindinglabel/apibindinglabel_controller.go

```diff
@@ -0,0 +1,247 @@
+
+	logging.WithQueueKey(logger, key).V(2).Info(fmt.Sprintf("queueing APIBinding%s", logSuffix))
+	c.queue.Add(key)
+}
+
+// enqueueAPIExport enqueues maps an APIExport to APIBindings for enqueuing.
+func (c *controller) enqueueAPIExport(obj interface{}, logger logr.Logger) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	bindingsForExport, err := c.getAPIBindingsByAPIExportKey(key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, binding := range bindingsForExport {
+		c.enqueueAPIBinding(binding, logging.WithObject(logger, obj.(*apisv1alpha1.APIExport)), " because of APIExport")
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := logging.WithReconciler(klog.FromContext(ctx), ControllerName)
+	ctx = klog.NewContext(ctx, logger)
+	logger.Info("Starting controller")
+	defer logger.Info("Shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	logger := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, logger)
+	logger.V(1).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *controller) process(ctx context.Context, key string) error {
+	logger := klog.FromContext(ctx)
+	clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		logger.Error(err, "invalid key")
+		return nil
+	}
+	apiBinding, err := c.apiBindingsLister.Cluster(clusterName).Get(name)
+	if errors.IsNotFound(err) {
+		return nil // object deleted before we handled it
+	}
+	if err != nil {
+		return err
+	}
+
+	if apiBinding.Spec.Reference.Workspace == nil {
+		return nil
+	}
+
+	apiExportClusterName := logicalcluster.New(apiBinding.Spec.Reference.Workspace.Path)
+	apiExport, err := c.apiExportsLister.Cluster(apiExportClusterName).Get(apiBinding.Spec.Reference.Workspace.ExportName)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	labels := map[string]interface{}{} // nil means to remove the key
+	var apiExportHasLabel, apiBindingHasLabel bool
+	if len(apiExport.Labels) > 0 {
```

Comment:
Is this test necessary ? retrieval from a nil map just works (as if the map doesn't contain anything), isn't it ?